Blog: Building Jamstack Infrastructure With Pulumi (#3530)
pulumi/docs: content/blog/jamstack-with-pulumi/index.md

meta_image: jamstack.png
authors:
    - sophia-parafina
tags:
    - jamstack
    - aws
---

A Jamstack is a modern architecture for building websites; JAM stands for JavaScript, APIs, and Markup. Jamstacks are deployed on a [CDN](https://en.wikipedia.org/wiki/Content_delivery_network), and content is stored on a cloud services provider. In addition to the speed and simplicity of deploying static content served from a CDN, there are other advantages such as maintaining content with git, modern build tools to generate the static content, automated builds, atomic deploys, and instant cache validation.

While build tools have simplified the process of creating content ready for deployment on a CDN, creating the infrastructure to serve the content remains complicated. . You can use a cloud provider’s web interface or script the build using a CLI tool if you want to manage your infrastructure instead of using a hosted solution. The alternative is to use infrastructure as code tool to automate building and deploying cloud resources. This article demonstrates how to create a jamstack website and deploy it on AWS using Pulumi.

<!--more-->

## Creating Your Website with Hugo

> There are many static site generators, but we’ll use hugo to build the site for this example. Instead of installing hugo, we’ll use [Docker](https://docs.docker.com/get-docker/) to run hugo in a [container](https://hub.docker.com/r/klakegg/hugo/). If you have Docker installed, create a directory to hold the website and change into that directory.

Reviewer: ```suggestion
There are many static site generators, but we’ll use [Hugo](https://gohugo.io) to build the site for this example. Instead of installing Hugo, we’ll use [Docker](https://docs.docker.com/get-docker/) to run Hugo in a [container](https://hub.docker.com/r/klakegg/hugo/). If you have Docker installed, create a directory to hold the website and change into that directory.
```

> There are many static site generators, but we’ll use hugo to build the site for this example. Instead of installing hugo, we’ll use [Docker](https://docs.docker.com/get-docker/) to run hugo in a [container](https://hub.docker.com/r/klakegg/hugo/). If you have Docker installed, create a directory to hold the website and change into that directory.

Reviewer 2: ```suggestion
There are many static site generators, but we’ll use Hugo to build the site for this example. We’ll use [Docker](https://docs.docker.com/get-docker/) to run Hugo in a [container](https://hub.docker.com/r/klakegg/hugo/). If you have Docker installed, create a directory to hold the website and change into that directory.
```

```bash
$ mkdir hugo
$ cd hugo
$ docker run --rm -it -v $(pwd):/src   klakegg/hugo:0.72.0-alpine shell
```

The `docker run` command will fetch the container (klakegg/hugo:0.72.0-alpine), map the current directory to the `src` directory in the container, and open a shell.

> In the container command line, we will use hugo to create a new site called *platypus*.

Reviewer: ```suggestion
In the container command line, we will use `hugo` to create a new site called *platypus*.
```

> In the container command line, we will use hugo to create a new site called *platypus*.

Reviewer 2: ```suggestion
In the container command line, we will use Hugo to create a new site called *platypus*.
```

```bash
$ hugo new site platypus
$ cd platypus
```

Next, we’ll customize the site with a theme. There are many [themes available](https://themes.gohugo.io/), but for this example, we'll use the *ananke* theme. Download the theme to the container, unzip it, and rename the unzipped directory to *ananke*.

```bash
$ wget https://github.com/budparr/gohugo-theme-ananke/archive/master.zip
$ unzip master.zip
$ cp -rf gohugo-theme-ananke-master ./themes/ananke
$ rm -rf master.zip gohugo-theme-ananke-master/
```

Configure the site by editing the *config.toml* file with the following.

```toml
baseURL = "http://www.<your_domain>.com/"
languageCode = "en-us"
title = "Platypus"
theme = "ananke"
```

> Use hugo to create a new markdown page in the posts directory.

Reviewer: ```suggestion
Use `hugo` to create a new markdown page in the posts directory.
```

```bash
$ hugo new posts/platypus.md
```

Now it’s time to add content to your site. You can use your favorite editor to write and edit content in your working directory on your machine instead of using the container shell. Edit the markdown file, below is an example page but feel free to add your content.

```md
## Platypus of the Day

**This platypus is a cool critter!**

![platypus](https://static01.nyt.com/images/2017/08/01/science/29TB-PLATYPUS1/29TB-PLATYPUS1-superJumbo.jpg)
```

> Finally, we can build the static site with hugo in the container command line.

Reviewer: ```suggestion
Finally, we can build the static site with `hugo` in the container command line.
```

> Finally, we can build the static site with hugo in the container command line.

Reviewer 2: ```suggestion
Finally, we can build the static site with Hugo in the container command line.
```

```bash
$ hugo -D
```

Hugo will create a *public* directory with the static website.

## Building the Cloud Infrastructure

Configuring a static website can be complicated when using either the AWS web interfaces or the AWS CLI. To deploy a static website, we need to create and configure the following cloud resources.

- [Amazon S3](https://aws.amazon.com/s3/) to store the website's contents
- [Amazon CloudFront](https://aws.amazon.com/cloudfront/) for the CDN serving content
- [Amazon Route53](https://aws.amazon.com/route53/) to set up the DNS for the website
- [Amazon Certificate Manager](https://aws.amazon.com/certificate-manager/) for securing the site via HTTPS

> Using Pulumi to deploy Infrastructure as Code, we can create and build our infrastructure using Python or your favorite programming language. If you haven't installed Pulumi and configured it to work with your AWS credentials, follow the [Getting Started with AWS guide]({{< relref "/docs/get-started/aws">}}).

Reviewer 3: Might also be worth mentioning that since this example uses Python, make sure to choose Python as your language pref.

> Using Pulumi to deploy Infrastructure as Code, we can create and build our infrastructure using Python or your favorite programming language. If you haven't installed Pulumi and configured it to work with your AWS credentials, follow the [Getting Started with AWS guide]({{< relref "/docs/get-started/aws">}}).

Reviewer: Nit:

```suggestion
Using Pulumi to deploy Infrastructure as Code, we can create and build our infrastructure using Python or your favorite programming language. If you haven't installed Pulumi and configured it to work with your AWS credentials, follow the [Getting Started with AWS guide]({{< relref "/docs/get-started/aws" >}}).
```

> To get started building our infrastructure, we’ll download the Python example for setting up a static website with AWS. There are many [examples on our GitHub repository](https://github.com/pulumi/), but we can clone just the [AWS Python Static Website example](https://github.com/pulumi/examples/tree/master/aws-py-static-website) using a sparse clone:

Reviewer 3: It might be good to explain what a sparse clone is, or link to something for it (e.g, I didn't immediately know what these commands were doing).

Author: added link

```bash

> $ mkdir aws-website

Reviewer 3: We should probably mention where this `mkdir` should be run, in relation to the Hugo site root, and that (presumably) we're no longer in the container here.

$ cd aws-website

$ git init
$ git remote add origin -f https://github.com/pulumi/examples/
$ git config core.sparseCheckout true
$ echo aws-py-static-website >> .git/info/sparse-checkout
$ git pull origin master
```

## Configuring and Deploying the Website

Now that we have a project to build the website, we will need to configure it so that it uses your domain name. Here are the steps to get it running.

1. Create a new stack or instance for testing:

    ```bash

>     $ pulumi stack init website-testing

Reviewer 3: When I run this command in the directory I was in when I left off from the last step above, I get an error about there not being a Pulumi project. We should probably add a `cd aws-py-static-website`.

    ```

2. Set the AWS region:

    ```bash

>     $ pulumi config set aws:region us-east-1

Reviewer 3: Does this have to be `us-east-1`, or can it be some other region? Might be good to call it out either way.

Author: Added note that it can be any zone

    ```

3. Create a Python virtualenv, activate it, and install dependencies:

>     This installs the dependent packages [needed](https://www.pulumi.com/docs/intro/concepts/how-pulumi-works/) for our Pulumi program.

Reviewer: ```suggestion
    This installs the dependent packages [needed]({{< relref "/docs/intro/concepts/how-pulumi-works" >}}) for our Pulumi program.
```

    ```bash
    $ python3 -m venv venv
    $ source venv/bin/activate

>     $ pip3 install -r requirements.txt

Reviewer: This post sort of assumes the reader will be using bash (or some other Unix shell). Should we show the commands for Windows?

```batch
> python -m venv venv
> venv\Scripts\activate
> pip3 install -r requirements.txt
```

Reviewer 3: If you wanted to do this, you could use a [`pulumi-chooser` with the `os` type](https://github.com/pulumi/docs/tree/master/components#usage).

    ```

4. Edit `Pulumi.website-testing.yaml` to set the configuration parameters. The first parameter is the `targetDomain`, which is the domain for the website (e.g., www.example.com).  The parent domain must be a [Route53 Hosted Zone](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/migrate-dns-domain-in-use.html) in the same AWS account as where the Pulumi program is running. The second parameter is `pathToWebsiteContents`, the relative path to the website’s contents created in Hugo.

>     ```yaml
>     config:
>     aws:region: us-east-1
>     static-website:targetDomain: www.<your_domain>.com
>     static-website:pathToWebsiteContents: public
>     ```

Reviewer: Personally, I'd encourage folks to use `pulumi config set` rather than manually editing the stack's yaml file.

As it's written here, I don't think the yaml content will even work correctly since the stack is named `website-testing` and these config values are prefixed with `static-website` (they'd need to be prefixed with the stack name, `website-testing`).

```suggestion
    ```bash
    $ pulumi config set targetDomain www.<your_domain>.com
    $ pulumi config set pathToWebsiteContents public
    ```
```

5. The example copies the site content from a directory in the same directory as the Pulumi project. To keep this example simple, copy the `public` directory in your hugo installation to the `aws-py-static-website` directory.

    ```bash

>     $ cp -rf <path_to>/hugo/public <path_to>/aws-py-static-website

Reviewer 3: I think this would be under `platypus` as well. It might also be clearer to describe what the folder structure should be -- say, 

```
./hugo
|_ playtpus
|_ aws-aws-website
```

.. and then your `pulumi config set pathToWebsiteContents` could be `../platypus/public`, removing the need to copy the files.

    ```

> 6. Run `pulumi up` to preview and deploy changes.  After the preview is displayed, you will be prompted if you want to continue and deploy the website. When all the resources have been created, you will see a listing of the resources similar to this:

Reviewer 3: Worked on the first try! 🎉  (I happened to have a domain in Route 53 already.)

```bash

> Updating (website-testing):

Reviewer 2: maybe elide or stick in scrolling window?

 +   ├─ aws:s3:Bucket                  requestLogs                                        created
 +   ├─ aws:route53:Record             www.sophiaparafina.com-validation                  created
 +   ├─ aws:acm:CertificateValidation  certificateValidation                              created
 +   ├─ aws:cloudfront:Distribution    cdn                                                created
 +   └─ aws:route53:Record             www.sophiaparafina.com                             created

Outputs:
    cloudfront_domain              : "d1l27ffvw5kc32.cloudfront.net"
    content_bucket_url             : "s3://www.your_domain.com"
    content_bucket_website_endpoint: "www.your_domain.com.s3-website-us-west-2.amazonaws.com"
    target_domain_endpoint         : "https://www.your_domain.com/"

Resources:
    + 27 created

Duration: 4m15s

Permalink: https://app.pulumi.com/spara/static-website/website-testing/updates/12
```

The website has been deployed, and you can browse it with your domain URL. Since this is a test deployment, don't forget to shut down and delete the cloud resources by running `pulumi destroy`.

## Examine the Code

> Let's take a look at how the cloud resources are created, configured, and populated. First up is the S3 bucket where the static website is store.

Reviewer: ```suggestion
Let's take a look at how the cloud resources are created, configured, and populated. First up is the S3 bucket where the static website is stored.
```

```python
# Create an S3 bucket configured as a website bucket.
content_bucket = pulumi_aws.s3.Bucket('contentBucket',
    bucket=target_domain,
    acl='public-read',
    website={
        'index_document': 'index.html',
        'error_document': '404.html'
    })
```

We add the website content by crawling the `public` directory and converting the files to S3 objects.

```python
def crawl_directory(content_dir, f):
    """
    Crawl `content_dir` (including subdirectories) and apply the function `f` to each file.
    """
    for file in os.listdir(content_dir):
        filepath = os.path.join(content_dir, file)

        if os.path.isdir(filepath):
            crawl_directory(filepath, f)
        elif os.path.isfile(filepath):
            f(filepath)

web_contents_root_path = os.path.join(os.getcwd(), path_to_website_contents)
def bucket_object_converter(filepath):
    """

>     Takes a file path and returns an bucket object managed by Pulumi

Reviewer 3: ```suggestion
    Takes a file path and returns a bucket object managed by Pulumi
```

    """
    relative_path = filepath.replace(web_contents_root_path + '/', '')
    # Determine the mimetype using the `mimetypes` module.
    mime_type, _ = mimetypes.guess_type(filepath)
    content_file = pulumi_aws.s3.BucketObject(
        relative_path,
        key=relative_path,
        acl='public-read',
        bucket=content_bucket,
        content_type=mime_type,
        source=FileAsset(filepath),
        opts=ResourceOptions(parent=content_bucket)
    )

# Crawl the web content root path and convert the file paths to S3 object resources.
crawl_directory(web_contents_root_path, bucket_object_converter)
```

> Now that we have our content in an S3 bucket, we turn to configure and create the CDN that serves the website. The first task is to create an SSL certificate based on the domain name hosted on Route 53 DNS if we didn't add one in the Pulumi.website-testing.yaml configuration file.

Reviewer: ```suggestion
Now that we have our content in an S3 bucket, we turn to configure and create the CDN that serves the website. The first task is to create an SSL/TLS certificate based on the domain name hosted on Route 53 DNS if we didn't specify the optional `certificateArn` config value of an existing certificate.
```

```python
if certificate_arn is None:
    # CloudFront is in us-east-1 and expects the ACM certificate to also be in us-east-1.
    # So, we create an east_region provider specifically for these operations.
    east_region = pulumi_aws.Provider('east', profile=pulumi_aws.config.profile, region='us-east-1')

    # Get a certificate for our website domain name.
    certificate = pulumi_aws.acm.Certificate('certificate',
        domain_name=target_domain, validation_method='DNS', opts=ResourceOptions(provider=east_region))

```

We also create a bucket to hold the CDN logs for the website.

```python
# Create a logs bucket for the CloudFront logs
logs_bucket = pulumi_aws.s3.Bucket('requestLogs', bucket=f'{target_domain}-logs', acl='private')
```

> Now that we have an SSL certificate and a S3 bucket to store logs, we can create the CDN. In the CDN resource definition, `origin` sets the S3 bucket as the content source, the domain name, and the ports for serving content. We can also set the cache_behavior, the price class, access restrictions, the logging configuration, and other parameters.

Reviewer: ```suggestion
Now that we have an SSL/TLS certificate and a S3 bucket to store logs, we can create the CDN. In the CDN resource definition, `origin` sets the S3 bucket as the content source, the domain name, and the ports for serving content. We can also set the cache_behavior, the price class, access restrictions, the logging configuration, and other parameters.
```

        'includeCookies': False,
        'prefix': f'${target_domain}/',
    },
    # CloudFront typically takes 15 minutes to fully deploy a new distribution.
    # Skip waiting for that to complete.
    wait_for_deployment=False)
```

To complete the deployment, we set the `alias_a_record` to point the CDN to our domain name in Route 53.

```python
def create_alias_record(target_domain, distribution):
    """
    Create a Route 53 Alias A record from the target domain name to the CloudFront distribution.
    """
    subdomain, parent_domain = get_domain_and_subdomain(target_domain)
    hzid = pulumi_aws.route53.get_zone(name=parent_domain).id
    return pulumi_aws.route53.Record(target_domain,
        name=subdomain,
        zone_id=hzid,
        type='A',
        aliases=[
            {
                'name': distribution.domain_name,
                'zoneId': distribution.hosted_zone_id,
                'evaluateTargetHealth': True
            }
        ]
    )

alias_a_record = create_alias_record(target_domain, cdn)
```

In a single Python script, we've created all the cloud resources to deploy our content and served it via a CDN.

## Conclusion

> Building modern websites is evolving from provider-managed Content Management Systems or user-managed [LAMP](https://en.wikipedia.org/wiki/LAMP_(software_bundle)) stacks. Static websites deployed in the cloud are popular because they separate content generation from the infrastructure that serves content. However, building and maintaining cloud infrastructure can be complicated and require multiple steps when using web client for configuration. A website using inexpensive storage such as S3 and serving content via Cloudfront can be deployed, versioned, and maintained using your favorite programming language. The Pulumi [examples repository](https://github.com/pulumi/examples) has many more code samples that demonstrate how to build and deploy cloud infrastructure across different cloud providers, using your favorite programming language.

Reviewer: Would it also be worth adding a sentence that explicitly mentions "for the same example in TypeScript, see https://github.com/pulumi/examples/tree/master/aws-ts-static-website"?

Author: Added in the TypeScript examples with language chooser

> Learn how to automate your cloud infrastructure with Pulumi's [Getting Started tutorials]({{< relref "/docs/get-started">}}) and [User Guides]({{< relref "/docs/guides">}}), or join us on [Slack](https://slack.pulumi.com/) if you have questions.

Reviewer: Nit:

```suggestion
Learn how to automate your cloud infrastructure with Pulumi's [Getting Started tutorials]({{< relref "/docs/get-started" >}}) and [User Guides]({{< relref "/docs/guides" >}}), or join us on [Slack](https://slack.pulumi.com/) if you have questions.
```
